### ISSUE TITLE
Many functions not chainable because of f(i,k) 1st arg

### ISSUE THREAD

**Issue by drauschenbach** [Yonaba/Moses#45]
In lodash/underscore, most functions are chainable because function callbacks are passed a value. That allows constructs such as this in JavaScript:

```javascript
local positives = _.map({1,-2,3,-4}, math.abs)
```

Such constructs don't work in Moses because some (but not all) functions are passed in index argument before a key argument. This is essential for efficient Lua, but undesirable when function chaining is desired. Here's what you currently have to do in Moses:

```lua
positives = moses.map({1,-2,3,-4}, function(i,v) return math.abs(v) end)
```

And because of this, Moses has great function chaining functions like `pipe()` and `compose()` that it seems like I never have a chance to use.

Off the top of my head, one idea is to create a new method with an "i" suffix when a callback function will be passed an index before a key. Eg:

```lua
positives = moses.mapi({1,-2,3,-4}, function(i,v) return math.abs(v) end)
-- or
positives = moses.map({1,-2,3,-4}, math.abs) 
```

**Comment by Yonaba** [comment]
This is something I was aware of, and I am planning working on this.
I was not considering adding new methods with "i" suffixes, as it would clunk too much the actual code, with too much duplicates, in my opinion. Instead I was considering reworking the whole library itself, so that constructs will accept/look for values first, indices next. 

This sounds like a viable option, but will need a lot of work, and will introduce breaking changes.

**Comment by drauschenbach** [comment]
Breaking changes are OK, and thanks to [SemVer](http://semver.org), we have an easy way to communicate them automatically to users and dependent modules.

**Comment by RussellHaley** [comment]
Seems like a lot of refactoring so someone doesn't have to do this:

```
local function f(inner)
	return function (i,v) return inner(v) end
end
positives = m.map({1,-2,3,-4}, f(math.abs))
positives = m.map({1,-2,3,-4}, f(math.deg))
positives = m.map({1,-2,3,-4}, f(math.ceil))
```

**Comment by Yonaba** [comment]
This have been fixed with commit 9bf8520.
